.gitlab-ci.yml

```diff
@@ -13,35 +13,61 @@ services:
 # If cache is defined outside the scope of jobs, it means it is set globally and all jobs will use that definition
 cache:
   paths:
+    - dist
+    - build
 
 # before_script is used to define the command that should be run before all jobs, including deploy jobs, but after the restoration of artifacts.
 # This can be an array or a multi-line string.
 before_script:
   - docker login -u $DOCKER_REGISTRY_USER_LOGIN -p $CI_REGISTRY_PASS_LOGIN $CI_NX_REGISTRY
-  - pip install twine
 
 # The YAML file defines a set of jobs with constraints stating when they should be run.
 # You can specify an unlimited number of jobs which are defined as top-level elements with an arbitrary name and always have to contain at least the script clause.
 # In this case we have only the test job which produce an artifacts (it must be placed into a directory called "public")
 # It is also specified that only the master branch will be subject of this job.
 
 stages:
+  - build
+  - test
```

Comment:
Is there a reason why we build the artifact before we run the tests?

Reply:
Good point. I also thought it should be the other way around, but this is the practice used by the Gitlab documentation itself. I couldn't find a rationale for it though, it's just what they do, so perhaps we could revert it.

Reply:
One principle we want to maintain is that we promote the same build through different stages. This gives us reassurance about the package being exactly the same. In this spirit, packaging could be the first action, so that if package is installed and tested correctly, it gets promoted without being rebuilt.